src/Validation.hs

```diff
@@ -0,0 +1,395 @@
+        Failure e' -> e <> e'
+        Success _  -> e
+    Success _ *> Failure e  = Failure e
+    Success _ *> Success b  = Success b
+    {-# INLINE (*>) #-}
+
+    (<*) :: Validation e a -> Validation e b -> Validation e a
+    Failure e <* b = Failure $ case b of
+        Failure e' -> e <> e'
+        Success _  -> e
+    Success _ <* Failure e  = Failure e
+    Success a <* Success _  = Success a
+    {-# INLINE (<*) #-}
+
+
+{- | This instance implements the following behavior for the binary operator:
+
+1. Both 'Failure': combine values inside 'Failure' using '<>'.
+2. At least is 'Success': return the left 'Success' (the earliest 'Success').
+3. 'empty' is @'Failure' 'mempty'@.
+
+__Examples__
+
+>>> success1 = Success [42] :: Validation [Text] [Int]
+>>> success2 = Success [69] :: Validation [Text] [Int]
+>>> failure1 = Failure ["WRONG"] :: Validation [Text] [Int]
+>>> failure2 = Failure ["FAIL"]  :: Validation [Text] [Int]
+
+>>> success1 <|> success2
+Success [42]
+>>> failure1 <|> failure2
+Failure ["WRONG","FAIL"]
+>>> failure2 <|> success2
+Success [69]
+-}
+instance (Semigroup e, Monoid e) => Alternative (Validation e) where
+    empty :: Validation e a
+    empty = Failure mempty
+    {-# INLINE empty #-}
+
+    (<|>) :: Validation e a -> Validation e a -> Validation e a
+    s@Success{} <|> _ = s
+    _ <|> s@Success{} = s
+    Failure e <|> Failure e' = Failure (e <> e')
+    {-# INLINE (<|>) #-}
+
+instance Foldable (Validation e) where
+    fold :: Monoid m => Validation e m -> m
+    fold (Success a) = a
+    fold (Failure _) = mempty
+    {-# INLINE fold #-}
+
+    foldMap :: Monoid m => (a -> m) -> Validation e a -> m
+    foldMap _ (Failure _) = mempty
+    foldMap f (Success a) = f a
+    {-# INLINE foldMap #-}
+
+    foldr :: (a -> b -> b) -> b -> Validation e a -> b
+    foldr f x (Success a) = f a x
+    foldr _ x (Failure _) = x
+    {-# INLINE foldr #-}
+
+instance Traversable (Validation e) where
+    traverse :: Applicative f => (a -> f b) -> Validation e a -> f (Validation e b)
+    traverse f (Success a) = Success <$> f a
+    traverse _ (Failure e) = pure (Failure e)
+    {-# INLINE traverse #-}
+
+    sequenceA :: Applicative f => Validation e (f a) -> f (Validation e a)
+    sequenceA = traverse id
+    {-# INLINE sequenceA #-}
+
+instance Bifunctor Validation where
+    bimap :: (e -> d) -> (a -> b) -> Validation e a -> Validation d b
+    bimap f _ (Failure e) = Failure (f e)
+    bimap _ g (Success a) = Success (g a)
+    {-# INLINE bimap #-}
+
+    first :: (e -> d) -> Validation e a -> Validation d a
+    first f (Failure e) = Failure (f e)
+    first _ (Success a) = Success a
+    {-# INLINE first #-}
+
+    second :: (a -> b) -> Validation e a -> Validation e b
+    second _ (Failure e) = Failure e
+    second g (Success a) = Success (g a)
+    {-# INLINE second #-}
+
+#if MIN_VERSION_base(4,10,0)
+instance Bifoldable Validation where
+--    bifoldMap :: (e -> m) -> (a -> m) -> Validation e a -> m
```

Comment:
Interesting...

Reply:
There was a problem with interaction between `-Wredundant-constraints` and `InstanceSigs`. Probably should submit GHC bug...